### DISCUSSION TITLE
ILGPU - Processing Data on your GPU (CUDA / OpenCL in C#)

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#3] [General]
# ILGPU - Processing Data on your GPU (CUDA / OpenCL in C#)

If you ever needed a massive amount of data processed in parallel you might had a look at OpenCL or CUDA.
Both a based on a subset of C to describe the program your GPU has to execute.
So let's have a look how we can achieve this in the .NET way. Furthermore I'll give you some insights about GPGPU (General-purpose computing on graphics processing units) in general.
Also some limitations and big differences to conventional "CPU"-Programming.

https://steven-giesel.com/blogPost/9e6b1539-7885-4288-ad39-4ba860f9e8fb

**Comment by StephanMoeller** [comment]
The version "1.0.0-rc2" is not available. Sitting in vs 2022, running this command:

Install-Package ILGPU -Version 1.0.0-rc2

And getting this error:

PM> Install-Package ILGPU -Version 1.0.0-rc2
Restoring packages for C:\git\ILGPU\ILGPU\ILGPU.csproj...
Install-Package : NU1108: Cycle detected. 
 ILGPU -> ILGPU (>= 1.0.0-rc2).
At line:1 char:1
+ Install-Package ILGPU -Version 1.0.0-rc2
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  + CategoryInfo     : NotSpecified: (:) [Install-Package], Exception
  + FullyQualifiedErrorId : NuGetCmdletUnhandledException,NuGet.PackageManagement.PowerShellCmdlets.InstallPackageCommand
 
Install-Package : Package restore failed. Rolling back package changes for 'ILGPU'.
At line:1 char:1
+ Install-Package ILGPU -Version 1.0.0-rc2
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  + CategoryInfo     : NotSpecified: (:) [Install-Package], Exception
  + FullyQualifiedErrorId : NuGetCmdletUnhandledException,NuGet.PackageManagement.PowerShellCmdlets.InstallPackageCommand
 

Any idea what I am doing wrong?

Thanks :)

**Reply by StephanMoeller** [reply]
Same issue with dotnet add:

PS C:\git\ILGPU\ILGPU> dotnet add package ILGPU --version 1.0.0-rc2

Welcome to .NET 6.0!
---------------------
SDK Version: 6.0.100-rc.2.21505.57

Telemetry
---------
The .NET tools collect usage data in order to help us improve your experience. It is collected by Microsoft and shared with the community. You can opt-out of telemetry by setting the DOTNET_CLI_TELEMETRY_OPTOUT environment variable to '1' or 'true' using your favorite shell.

Read more about .NET CLI Tools telemetry: https://aka.ms/dotnet-cli-telemetry

----------------
Installed an ASP.NET Core HTTPS development certificate.
To trust the certificate run 'dotnet dev-certs https --trust' (Windows and macOS only).
Learn about HTTPS: https://aka.ms/dotnet-https
----------------
Write your first app: https://aka.ms/dotnet-hello-world
Find out what's new: https://aka.ms/dotnet-whats-new
Explore documentation: https://aka.ms/dotnet-docs
Report issues and find source on GitHub: https://github.com/dotnet/core
Use 'dotnet --help' to see available commands or visit: https://aka.ms/dotnet-cli
--------------------------------------------------------------------------------------
  Determining projects to restore...
  Writing C:\Users\Stephan\AppData\Local\Temp\tmpF51C.tmp
info : Adding PackageReference for package 'ILGPU' into project 'C:\git\ILGPU\ILGPU\ILGPU.csproj'.
info : Restoring packages for C:\git\ILGPU\ILGPU\ILGPU.csproj...
error: NU1108: Cycle detected.
error:   ILGPU -> ILGPU (>= 1.0.0-rc2).
info : Package 'ILGPU' is compatible with all the specified frameworks in project 'C:\git\ILGPU\ILGPU\ILGPU.csproj'.
error: Value cannot be null. (Parameter 'path1')

**Reply by StephanMoeller** [reply]
It works with VS2019 and dotnet 5 projects. Ill just use that then :)

**Reply by linkdotnet** [reply]
Hey @SRYER.

Originally I used .net5 for the given example in the blog post. Therefore I setup a new project with .net 6 rc2 and it works as you would expect.

However the reason why it isn't working could be funnier than you think: In your output from nuget you showed
> error: NU1108: Cycle detected.
> error: ILGPU -> ILGPU (>= 1.0.0-rc2).

This can happen because your project has the same name as the nuget package. Could you try to rename the project?
Here more information: https://kmatyaszek.github.io/2019/05/30/nuget-nu1108-cycle-detected.html
